modules/navier_stokes/src/problems/NavierStokesProblem.C

```diff
@@ -0,0 +1,344 @@
+                                intermediate_is,
+                                intermediate_is,
+                                MAT_INITIAL_MATRIX,
+                                &intermediate_mat);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+      intermediate_mats[i] = intermediate_mat;
+    }
+    return _index_sets.empty() ? parent_mat : intermediate_mats.back();
+  };
+
+  Mat our_parent_Q = nullptr;
+  if (_have_mass_matrix)
+    our_parent_Q = process_intermediate_mats(intermediate_Qs, global_Q);
+  Mat our_parent_L = nullptr;
+  if (_have_L_matrix)
+    our_parent_L = process_intermediate_mats(intermediate_Ls, global_L);
+
+  // Need to call this before getting index sets or sub ksps, etc.
+  ierr = PCSetUp(schur_pc);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+
+  ierr = PCFieldSplitGetISByIndex(schur_pc, 0, &velocity_is);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+  ierr = MatGetOwnershipRange(our_parent_Q, &rstart, &rend);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+
+  if (_commute_lsc)
+  {
+    mooseAssert(our_parent_L, "This should be non-null");
+    if (!_L)
+    {
+      ierr = MatCreateSubMatrix(our_parent_L, velocity_is, velocity_is, MAT_INITIAL_MATRIX, &_L);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+    else
+    {
+      ierr = MatCreateSubMatrix(our_parent_L, velocity_is, velocity_is, MAT_REUSE_MATRIX, &_L);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+  }
+
+  ierr = ISComplement(velocity_is, rstart, rend, &pressure_is);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+  if (!_Q_scale)
+  {
+    if (_commute_lsc || _pressure_mass_matrix_as_pre)
+    {
+      mooseAssert(our_parent_Q, "This should be non-null");
+      ierr =
+          MatCreateSubMatrix(our_parent_Q, pressure_is, pressure_is, MAT_INITIAL_MATRIX, &_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+    else if (_have_mass_matrix) // If we don't have a mass matrix and the user has requested scaling
+                                // then the diagonal of A will be used
+    {
+      mooseAssert(our_parent_Q, "This should be non-null");
+      ierr =
+          MatCreateSubMatrix(our_parent_Q, velocity_is, velocity_is, MAT_INITIAL_MATRIX, &_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+  }
+  else
+  {
+    mooseAssert(our_parent_Q, "This should be non-null");
+    if (_commute_lsc || _pressure_mass_matrix_as_pre)
+    {
+      ierr =
+          MatCreateSubMatrix(our_parent_Q, pressure_is, pressure_is, MAT_REUSE_MATRIX, &_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+    else
+    {
+      ierr =
+          MatCreateSubMatrix(our_parent_Q, velocity_is, velocity_is, MAT_REUSE_MATRIX, &_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+  }
+  ierr = ISDestroy(&pressure_is);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+
+  for (auto & mat : intermediate_Qs)
+  {
+    ierr = MatDestroy(&mat);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+  }
+  for (auto & mat : intermediate_Ls)
```

Comment:
why do we store them if we just end up destroy them here?

Reply:
It's just an easy way to progress to the sub-matrix representing the velocity-pressure dofs we are going to do LSC for. I could do more of a recursion that only ever leads to storage of one parent sub-matrix